src/lib.rs

```diff
@@ -196,21 +219,81 @@ fn do_camera_movement(
     let delta_device_pixels = current_pos - last_pos.unwrap_or(current_pos);
 
     for (cam, mut transform, projection) in &mut query {
-        if !cam.enabled
-            || !cam
-                .grab_buttons
-                .iter()
-                .any(|btn| mouse_buttons.pressed(*btn) && !mouse_buttons.just_pressed(*btn))
-        {
-            continue;
-        }
-
-        let proj_area_size = projection.area.size();
-        let world_units_per_device_pixel = proj_area_size / window_size;
-
-        // The proposed new camera position
-        let delta_world = delta_device_pixels * world_units_per_device_pixel;
-        let proposed_cam_pos = transform.translation.truncate() - delta_world;
+        let (proposed_cam_pos, proj_area_size) = match cam.move_mode {
+            MoveMode::Mouse => {
+                if !cam.enabled
+                    || !cam
+                        .grab_buttons
+                        .iter()
+                        .any(|btn| mouse_buttons.pressed(*btn) && !mouse_buttons.just_pressed(*btn))
+                {
+                    continue;
+                }
+
+                let proj_area_size = projection.area.size();
+                let world_units_per_device_pixel = proj_area_size / window_size;
+
+                // The proposed new camera position
+                let delta_world = delta_device_pixels * world_units_per_device_pixel;
+                (
+                    transform.translation.truncate() - delta_world,
+                    proj_area_size,
+                )
+            }
+            MoveMode::Keyboard => {
+                let mut direction = Vec2::ZERO;
+
+                if cam
+                    .move_keys
+                    .left
+                    .iter()
+                    .any(|key| keyboard_buttons.pressed(*key))
+                {
+                    direction.x += 1.;
+                }
+
+                if cam
+                    .move_keys
+                    .right
+                    .iter()
+                    .any(|key| keyboard_buttons.pressed(*key))
+                {
+                    direction.x -= 1.;
+                }
+
+                if cam
+                    .move_keys
+                    .up
+                    .iter()
+                    .any(|key| keyboard_buttons.pressed(*key))
+                {
+                    direction.y -= 1.;
+                }
+
+                if cam
+                    .move_keys
+                    .down
+                    .iter()
+                    .any(|key| keyboard_buttons.pressed(*key))
+                {
+                    direction.y += 1.;
+                }
+
+                if direction.length() == 0. {
+                    continue;
+                }
```

Comment:
not sure why we need to have this. i removed it and it works. or if i'm misunderstanding the point, it would be faster to use `length_squared`

Reply:
Struggle a bit with reading this on my phone, but I'd assume avoiding triggering change detection when there is no actual change.

Reply:
Indeed that is why I put it there. I'll look into `length_squared` tomorrow.